### PR TITLE
Fix pad_last_batch in GPU NumpyReader

### PR DESCRIPTION
**Pull request by JanuszL**
- fixes a crash in the numpy reader when pad_last_batch is set and more
  then one thread is used by DALI
- when pad_last_batch is set the padding duplicates the last valid sample.
  In the case of numpy reader, it means duplication of structure with a file
  descriptor. As samples within the batch are read in parallel it leads to
  two threads reading the same file with the same file descriptor which causes
  unexpected changes in the offset inside the file during reading. This patch
  adds logic that handles padded samples.

Signed-off-by: Janusz Lisiecki <jlisiecki@nvidia.com>

## Category:
**Bug fix** (*non-breaking change which fixes an issue*)

## Description:
- fixes a crash in the numpy reader when pad_last_batch is set and more
  then one thread is used by DALI
- when pad_last_batch is set the padding duplicates the last valid sample.
  In the case of numpy reader, it means duplication of structure with a file
  descriptor. As samples within the batch are read in parallel it leads to
  two threads reading the same file with the same file descriptor which causes
  unexpected changes in the offset inside the file during reading. This patch
  adds logic that handles padded samples.

### Affected modules and functionalities:
- GPU numpy reader

### Key points relevant for the review:
- logic

### Tests:

- [ ] Existing tests apply
- [x] New tests added
  - [x] Python tests
    - test_operator_readers_numpy.py:test_pad_last_sample
  - [ ] GTests
  - [ ] Benchmark
  - [ ] Other
- [ ] N/A

### Documentation
- [x] Existing documentation applies
- [ ] Documentation updated
  - [ ] Docstring
  - [ ] Doxygen
  - [ ] RST
  - [ ] Jupyter
  - [ ] Other
- [ ] N/A

#### Requirements
- [ ] Implements new requirements
- [ ] Affects existing requirements
- [x] N/A

**REQ IDs**: N/A

**JIRA TASK**: N/A

### CODE REVIEW

File: dali/operators/reader/numpy_reader_gpu_op.cc [hunk 1]
```diff
@@ -85,15 +88,30 @@ void NumpyReaderGPU::Prefetch() {
...
   staging_.commit();
+
+  // copy padded samples
+  for (int data_idx = first_padded; data_idx < curr_tensor_list.num_samples(); ++data_idx) {
+    kernels::copy<StorageGPU, StorageGPU>(curr_tensor_list.raw_mutable_tensor(data_idx),
+                                          curr_tensor_list.raw_tensor(data_idx - 1),
+                                          curr_tensor_list.tensor_offset(data_idx) -
+                                            curr_tensor_list.tensor_offset(data_idx - 1),
+                                          staging_stream_);
+  }
```
Comment (mzient) [inline dali/operators/reader/numpy_reader_gpu_op.cc:114]: I think this should rather be a scatter-gather operation (with overlapping source).

File: dali/operators/reader/numpy_reader_gpu_op.cc [hunk 2]
```diff
@@ -85,15 +88,30 @@ void NumpyReaderGPU::Prefetch() {
...
+      ScheduleChunkedRead(sample, *curr_batch[data_idx]);
+    }
   }
   thread_pool_.RunAll();
   staging_.commit();
+
+  // copy padded samples
+  for (int data_idx = first_padded; data_idx < curr_tensor_list.num_samples(); ++data_idx) {
+    kernels::copy<StorageGPU, StorageGPU>(curr_tensor_list.raw_mutable_tensor(data_idx),
+                                          curr_tensor_list.raw_tensor(data_idx - 1),
```
Comment (mzient) [inline dali/operators/reader/numpy_reader_gpu_op.cc:110]:
```suggestion
                                          curr_tensor_list.raw_tensor(first_padded - 1),
```

File: dali/operators/reader/numpy_reader_gpu_op.cc [hunk 3]
```diff
@@ -85,15 +88,30 @@ void NumpyReaderGPU::Prefetch() {
...
+                                          curr_tensor_list.tensor_offset(data_idx) -
+                                            curr_tensor_list.tensor_offset(data_idx - 1),
```
[the remark this answers is not in the document]
Comment (JanuszL) [reply dali/operators/reader/numpy_reader_gpu_op.cc:112]: Good catch.

File: dali/operators/reader/numpy_reader_gpu_op.cc (hunk 1 above)
Comment (JanuszL) [reply dali/operators/reader/numpy_reader_gpu_op.cc:114]: Will do.

**Comment by JanuszL** [comment]
!build

File: dali/operators/reader/loader/numpy_loader_gpu.h [hunk 4]
```diff
@@ -49,6 +49,7 @@ struct NumpyFileWrapperGPU {
   TensorShape<> shape;
   DALIDataType type;
   DALIMeta meta;
+  int source_sample_idx;
```
Comment (mzient) [inline dali/operators/reader/loader/numpy_loader_gpu.h:52]: I'd suggest initializing (it with something invalid)
```suggestion
  int source_sample_idx = -1;
```

File: dali/operators/reader/numpy_reader_gpu_op.cc [hunk 5]
```diff
@@ -85,12 +88,26 @@ void NumpyReaderGPU::Prefetch() {
...
+      }
+    } else {
+      ScheduleChunkedRead(sample, *curr_batch[data_idx]);
+    }
+    if (first_padded > 0) {
+      // connect padded samples with the original data idx
+      curr_batch[data_idx]->source_sample_idx = first_padded - 1;
+    } else {
+      curr_batch[data_idx]->source_sample_idx = data_idx;
+    }
```
Comment (mzient) [inline dali/operators/reader/numpy_reader_gpu_op.cc:110]:
```suggestion
```

File: dali/operators/reader/numpy_reader_gpu_op.cc [hunk 6]
```diff
@@ -85,12 +88,26 @@ void NumpyReaderGPU::Prefetch() {
...
     curr_tensor_list.SetMeta(data_idx, curr_batch[data_idx]->meta);
     SampleView<GPUBackend> sample(curr_tensor_list.raw_mutable_tensor(data_idx),
                                   curr_tensor_list.tensor_shape(data_idx),
                                   curr_tensor_list.type());
-    ScheduleChunkedRead(sample, *curr_batch[data_idx]);
+    // when padding, the last sample is duplicated so no need to redo the same work
+    if (data_idx > 0 && curr_batch[data_idx - 1] == curr_batch[data_idx]) {
+      if (first_padded < 0) {
+        first_padded = data_idx;
+      }
```
Comment (mzient) [inline dali/operators/reader/numpy_reader_gpu_op.cc:101]:
```suggestion
      }
      curr_batch[data_idx]->source_sample_idx = first_padded - 1;      
```

File: dali/operators/reader/numpy_reader_gpu_op.cc [hunk 7]
```diff
@@ -85,12 +88,26 @@ void NumpyReaderGPU::Prefetch() {
...
+    } else {
+      ScheduleChunkedRead(sample, *curr_batch[data_idx]);
```
Comment (mzient) [inline dali/operators/reader/numpy_reader_gpu_op.cc:103]:
```suggestion
      ScheduleChunkedRead(sample, *curr_batch[data_idx]);
      curr_batch[data_idx]->source_sample_idx = data_idx;
```

**Comment by JanuszL** [comment]
!build

File: dali/test/python/test_operator_readers_numpy.py [hunk 8]
```diff
@@ -572,3 +573,41 @@ def test_numpy_reader_roi_error():
...
+            arr_np_list.append(np.load(last_file_name))
+        pipe = NumpyReaderPipeline(path=test_data_root,
+                                files=filenames,
+                                file_list=None,
+                                file_filter=None,
+                                device=device,
+                                batch_size=batch_size,
+                                num_threads=4,
+                                device_id=0,
+                                pad_last_batch=True)
```
Comment (mzient) [inline dali/test/python/test_operator_readers_numpy.py:601]: PEP8 - align the indent to match the first argument.
```suggestion
        pipe = NumpyReaderPipeline(path=test_data_root,
                                   files=filenames,
                                   file_list=None,
                                   file_filter=None,
                                   device=device,
                                   batch_size=batch_size,
                                   num_threads=4,
                                   device_id=0,
                                   pad_last_batch=True)
```

**Comment by JanuszL** [comment]
!build

**Comment by JanuszL** [comment]
!build

**Review by mzient — APPROVED** [review]

**Review by stiepan — APPROVED** [review]

File: dali/operators/reader/numpy_reader_gpu_op_impl.cu [hunk 9]
```diff
@@ -53,18 +55,27 @@ void NumpyReaderGPU::RunImplTyped(DeviceWorkspace &ws) {
...
       nsamples_copy++;
+    source_data_index_[i] = GetSample(i).source_sample_idx;
   }
 
   if (nsamples_copy) {
-    if (nsamples_copy == nsamples) {
+    bool no_padded = true;
+    for (int i = 0; i < nsamples; ++i) {
+      if (source_data_index_[i] != i) {
+        no_padded = false;
```
[the remark this answers is not in the document]
Comment (mzient) [reply dali/operators/reader/numpy_reader_gpu_op_impl.cu:65]: Or perhaps something like `needs_copy` or, conversely `can_swap`.

File: dali/test/python/test_operator_readers_numpy.py (hunk 8 above)
Comment (JanuszL) [reply dali/test/python/test_operator_readers_numpy.py]: Done

File: dali/test/python/test_operator_readers_numpy.py [hunk 10]
```diff
@@ -572,3 +573,41 @@ def test_numpy_reader_roi_error():
...
+        filenames = []
+        arr_np_list = []
+        last_file_name = None
+        for index in range(0, num_samples):
+            filename = os.path.join(test_data_root, "test_{:02d}.npy".format(index))
+            last_file_name = filename
+            filenames.append(filename)
+            create_numpy_file(filename, (5, 2, 8), np.float32, False)
+            arr_np_list.append(np.load(filename))
+        for _ in range(num_samples, batch_size):
```
[the remark this answers is not in the document]
Comment (JanuszL) [reply dali/test/python/test_operator_readers_numpy.py]: Done

**Comment by JanuszL** [comment]
!build

File: dali/operators/reader/numpy_reader_gpu_op_impl.cu (hunk 9 above)
[the remark this answers is not in the document]
Comment (JanuszL) [reply dali/operators/reader/numpy_reader_gpu_op_impl.cu]: Done

**Pull request merged**